content/en/community/hackathons/sessions/behind-scenes/content/Work-Items_06.-Tutorial---Reminder:-Adding-Filesystems-to-an-Application-2023.md

```diff
@@ -0,0 +1,136 @@
+    * `Makefile` is used by the manual configuration and build system.
+    * `launch.sh` is a wrapper script around `qemu-system-x86_64` used to manually run the application.
+We'll use it as well to run the application.
+
+
+1. Clone the Unikraft repository:
+
+    ```console
+    $ mkdir .unikraft
+    $ cd .unikraft/
+    $ mkdir libs/
+    $ cd ../
+    ```
+
+1. Add the `lib-musl` repository:
+
+    For filesystem functionalities (opening, reading, writing files) we require a more powerful libc.
+    [Musl](https://github.com/unikraft/lib-musl) is already ported in Unikraft and will do nicely:
+
+    ```console
+    $ cd .unikraft/libs
+    $ git clone https://github.com/unikraft/lib-musl musl
+    ```
+
+    We also need to update the `LIBS` line in the `Makefile`:
+
+    ```console
+    LIBS := $(UK_LIBS)/musl
+    ```
+
+### Configuring the application
+
+Now we need to enable **9pfs** and **musl** in Unikraft.
+To do this, we run:
+
+```console
+$ make menuconfig
+```
+
+We need to select the following options, from the `Library Configuration` menu:
+
+* `libmusl`
+* `vfscore: VFS Core Interface`
+* `vfscore: VFS Configuration` -> `Automatically mount a root filesystem` -> `Default root filesystem` -> `9pfs`
+  * For the `Default root device` option fill the `fs0` string.
+
+These configurations will also enable the **9pfs** and **uk9p** internal libraries the menu.
+
+We want to run Unikraft with QEMU / KVM, so we must select **KVM guest** in the `Platform Configuration` menu.
+For 9pfs we also need to enable, in the **KVM guest** options menu, `Virtio` -> `Virtio PCI device support`.
+
+Save the configuration and exit.
+
+Do a quick check of the configuration in `.config` by pitting it against the `config.sol` file in the reference solution:
+
+```console
+$ diff -u .config ../../sol/03-adding-filesytstems/config.sol
+```
+Differences should be minimal, such as the application identifier.
+
+**Build**
+
+Build the Unikraft image:
+
+```console
+make
+```
+
+Building the Unikraft image will take a while.
+It has to pull musl source code, patch it and then build it, together with the Unikraft source code.
+
+**Run with qemu-system-x86_64**
+
+To run the Unikraft image with QEMU / KVM, we use the wrapper `launch.sh` script, that calls `qemu-system-x86_64` command with the proper arguments:
+
+```console
+$ ./launch.sh ./build/06-adding-filesystems_qemu-x86_64
+[...]
+Powered by
+o.   .o       _ _               __ _
+Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
+oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
+oOo oOO| | | | |   (| | | (_) |  _) :_
+ OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
+                  Atlas 0.13.1~5eb820bd
+Hello, world!
+File contents: The grass is green!
+Bye, world!
+```
+
+A completely manual run would use the command:
+
+```console
+$ qemu-system-x86_64 -fsdev local,id=myid,path=guest_fs,security_model=none -device virtio-9p-pci,fsdev=myid,mount_tag=fs0 -kernel build/06-adding-filesystems_qemu-x86_64 -nographic
```

Comment:
Somehow it hangs after "Hello, world!" message, and doesn't display the file content witout adding. If I add '-cpu host -enable-kvm' it runs as expected.

Reply:
People are working on virtual machines, we will confuse them with this. It hangs because of [this](https://github.com/unikraft/unikraft/pull/941). It's fine to leave it like this for now.